Initial implementation of lot architecture

riperiperi opened the pull request:
![image](https://github.com/LazyDuchess/OpenTS2/assets/6294155/846e656a-87d5-40f9-b72b-6099a1cc5bd6)

Adds support for:
- Terrain
  - Up to 32 blend textures, one draw. Creates a bitmask to select which blend textures to use.
  - Textures are placed on a texture array, but they can be different sizes... so they have to be blit into an uncompressed one.
- Floors
  - Diagonals, edges fully supported
- Walls
  - Intersections
  - Foundation/Decking
  - Fences (with slopes)
  - Roof-walls
  - Half-walls
  - Groundwork for object cutouts is in place, something just needs to apply them...
- Roofs
  - Fancy intersections which match the base game wherever possible.
  - Normal, Diagonal, Pagoda, Diagonal pagoda...
- Water
- Controls for quickly switching lot in Play mode (use Controller in editor)
- All components have been split into floors so they can be shown/hidden by user controls, and partially updated.
- View level and walls mode configurable in Play mode (use Controller in editor)

TODO in future:
- Ceilings
- Pools
- A lot more 3ARY resources (rooms, reserved tiles, ceilings, etc)
- Rooms (and cutaways)
- Special roofs (octogonal, cone, sphere, greenhouse)
- Wall intersections should have different tracking for wall/roofwall/half wall
- Anything to do with editing
  - "Dirty" tracking for floors, regenerating modified floors when a change is made
  - Wall runtime structure with fast lookup and editing tools, useful for simulator placement rules too
- Bump maps on terrain textures
- Improve materials stuff
- Improve roof cuts when not tile-compatible.

riperiperi commented:
![image](https://github.com/LazyDuchess/OpenTS2/assets/6294155/9b998333-a34c-45d0-bbe4-6b1adfe18b6b)

Fences also work on slopes. Since unity doesn't seem to support passing in a custom matrix to a transform, it has to build a skew out of three stacked transforms.

ammaraskar commented on Assets/Scripts/OpenTS2/Rendering/Materials/FloorMaterial.cs:
```diff
@@ -0,0 +1,28 @@
+
+        public override Material Parse(ScenegraphMaterialDefinitionAsset definition)
+        {
+            // Includes floorMaterialScaleU and floorMaterialScaleV, currently unhandled.
```
is this comment outdated?

riperiperi replied:
Yes.

riperiperi commented:
Added some basic support for Roofs. Right now it only covers the base game roofs.

![image](https://github.com/LazyDuchess/OpenTS2/assets/6294155/a5dcda3d-45eb-4af7-bc81-6f3e8262bcb6)

![image-1](https://github.com/LazyDuchess/OpenTS2/assets/6294155/e653a215-246b-451f-afb7-5094145281e6)

![image](https://github.com/LazyDuchess/OpenTS2/assets/6294155/04f55f30-dc9b-4baa-8e0d-5fb4211c5d8d)

![image-1](https://github.com/LazyDuchess/OpenTS2/assets/6294155/86aad170-128d-4584-a883-2b86fda76de1)

TS2 builds roofs out of quarter tiles, each can be cut from four cardinal directions. Each roof is made up of a number of faces, which I've represented with `RoofEdge`. Each edge has a tilemap which represents what graphic should be used for each tile, all of which are stored in a 5x5 texture atlas in the top texture.

Roofs always delete tiles that are lower than them. I've added a simple case that does this when all tile corners are covered, but the interesting part is when grid-aligned roofs intersect. These have specific behaviours which precisely cut into other roof tiles, which make for some really clean roof geometry. Given certain types of intersection, an additional intersection layer is added to the tile. To prevent z-fighting, intersections where two faces are coplanar pick and choose which graphic is shown based on a priority system.

The diagonal roofs will work in a similar way, but they are not allowed to calculate intersections with the regular roofs - only each other. These aren't implemented yet, but hopefully shouldn't be too difficult.

The special roofs (dome, cone, curved) don't have any fancy intersection rules or graphics. They still need to be implemented, likely with custom geometry generators for each.

riperiperi commented:
I've pushed a change to the lot testing controller that makes it a lot quicker to switch property without reloading all the game content. It can leave some game objects behind due to object scenegraphs failing, though. Should make it easier to hop around and find issues. It also now loads all expansion content.

https://github.com/LazyDuchess/OpenTS2/assets/6294155/7f4c0af9-7395-4d36-ba3a-6feb4ca60a38

When any code changes are made, all game content seems to unload... which is unfortunate.

riperiperi commented:
https://github.com/LazyDuchess/OpenTS2/assets/6294155/9bdfb2ba-57b1-419d-b84e-6e89acebf3fe

I think I've done enough to open the PR for review. There's a lot left to do, but this is getting big enough as is.

The next large step would probably be tools for editing or interaction from the simulator, such as checking wall adjacency for a tile. There are a bunch of smaller things I've listed on the PR body.

ammaraskar reviewed (approved):
Looks good to me with some small suggestions

riperiperi commented on Assets/Scripts/OpenTS2/Content/DBPF/CatalogFenceAsset.cs:
```diff
@@ -0,0 +1,48 @@
+
+namespace OpenTS2.Content.DBPF
+{
+    public static class CatalogFenceType
```
I don't really know what to do about this. In the case of Fence and Roof, it's always "fence" and "roof" respectively, so there's not much value in exposing the field. For objects it could be anything, but we don't know all the possible values and an enum parse would immediately fail if it were something unexpected. It would be used there to filter out wall and floor category items, but the only use right now is a GUID lookup that just believes the entry has the correct type. It should be used to populate the catalog in future.

ammaraskar replied:
I see, this is fine for now then.

riperiperi commented on Assets/Scripts/OpenTS2/Scenes/Lot/LotArchitectureMeshComponent.cs:
```diff
@@ -0,0 +1,158 @@
+            _indexBuilder.Add(baseVertex + c);
+        }
+
+        public void AddWindingRect(int baseVert)
```
Unused

ammaraskar commented on Assets/Scripts/OpenTS2/Content/DBPF/Scenegraph/ScenegraphMaterialDefinitionAsset.cs:
```diff
@@ -42,13 +42,13 @@ public override void FreeUnmanagedResources()
         }
 
-        public Material GetAsUnityMaterial()
+        public Material GetAsUnityMaterial(string forceType = null)
```
Aah I missed that file because it was collapsed by Github by default.

ammaraskar commented on Assets/Scripts/OpenTS2/Rendering/Materials/StandardMaterial.cs:
```diff
@@ -87,6 +97,13 @@ public override Material Parse(ScenegraphMaterialDefinitionAsset definition)
 
+            float alpha = alphaMul * untexturedAlpha;
+
+            if (alpha != 1)
```
I'd say just always set it, it's fine to have the default in multiple places.

ammaraskar commented on Assets/Scripts/OpenTS2/Files/Formats/DBPF/StringMapCodec.cs:
```diff
@@ -0,0 +1,60 @@
+                {
+                    Value = reader.ReadVariableLengthPascalString(),
+                    Id = reader.ReadUInt16(),
+                    Unknown1 = reader.ReadUInt16(),
```
Thanks for looking at the values, it's definitely a uint32 but I'm guessing it's not used anymore since the game definitely sets the value for any new entries of that field to 1. It does round-trip already existing values so I'm guessing those were already just saved like that.

The simplified reverse-engineered code for when it writes out a StringMap is:

```c
  (*stream->_vptr->SetUint32)(getClassId());
  (*stream->_vptr->SetUint32)(1);
  (*stream->_vptr->SetGZStr)("cStringMap");
  (*stream->_vptr->SetSint32)(this->numEntries);
  for (entry in this->entries) {
    (*stream->_vptr->SetGZStr)(entry->str);
    (*stream->_vptr->SetUint16)(entry->val);
    (*stream->_vptr->SetUint32)(entry->unknownInt32);
  }
```

LazyDuchess reviewed (approved):
All looks great to me!
